### ISSUE TITLE
Stop calling compute_form_data in ffc_interface.py

### ISSUE THREAD

**Issue by dorugeber** [firedrakeproject/firedrake#518]
`compute_form_data` is supposed to be called once per form, from inside FFC.

Currently, we call it a bunch of times in ffc_interface.py to get bits of information.  This was done due to the removal of form.form_data(), in a286c41 (undid one of these in b27934f).

One time is in the `__init__` function of the `FFCKernel` class; we examine the `elements` property of the returned data to see if any of them need orientations, then iterate over `preprocessed_form.integrals()` [this is *not* the original form; it has had derivatives expanded out, etc.]

The second time is just before we decide whether to split the form.  We look in `unique_sub_elements` to see if any of them are Mixed (this can be replaced by a check on `form.coefficients()` and `form.arguments()`  EDIT: actually just arguments, I hope).  We then store the `preprocessed_form.coefficients()` and `preprocessed_form.integrals()` along with the kernel.

The third time is similar to the second, but it is called on each sub-form if the form gets split.

**Comment by dorugeber** [comment]
Some comments from an email thread:

@wence-: For a given form, FFC spits back a list of kernels (e.g. over the different integral domains).  This has an ordering which comes from the preprocessed form.

To assemble the form, we need to have the kernel (obviously) + any arguments to the kernel (coefficients) but these need to come out in the same order.

That is, form.coefficients() may have a different order to preproc_form.coefficients() [at least I vaguely recall something along those line].

I notice that the FFCKernel constructor calls compute_form_data (at which point all info is available) and could construct the lists that are currently built in compile_form.  This would cut the calling down to once per form on the firedrake side (and presumably once inside ffc).  If the memoized values come back in the "canonical" order, we may be able to get away without the firedrake-side call at all.

Martin Alnaes: The relative ordering is always the same, but the preprocessed form may have fewer coefficients.

By the way, I did some improvements last year, take a look at

ufl/ufl/algorithms/compute_form_data.py : 240
    # Store back into integral data which form coefficients are used by each integral
    for itg_data in self.integral_data:
        itg_data.enabled_coefficients = [bool(coeff in itg_data.integral_coefficients)
                                         for coeff in self.reduced_coefficients]


dolfin/site-packages/dolfin/fem/form.py : 87
        # Extract coefficients from form (pass only the ones that
        # the compiled form actually uses to the assembler)
        original_coefficients = form.coefficients()
        self.coefficients = []
        for i in range(self._compiled_form.num_coefficients()):
            j = self._compiled_form.original_coefficient_position(i)
            self.coefficients.append(original_coefficients[j])

This information can be used to detect which coefficients you need to fetch local element dofs from for each integral.

**Comment by dorugeber** [comment]
Pfft, just tried replacing
```
if not any(type(e) is MixedElement for e in fd.unique_sub_elements):
```
by
```
if not any(type(a.element) is MixedElement for a in form.arguments()):
```
and pytest hard-quits a few tests into 2dcohomology.  Bleurgh.

**Comment by wence-** [comment]
Closed by #691
